src/realm/sync/client.cpp

```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+}
 
-void SessionWrapper::report_progress(bool only_if_new_uploadable_data)
+void SessionWrapper::report_progress(bool is_download, bool is_completed, bool only_if_new_uploadable_data)
 {
     REALM_ASSERT(!m_finalized);
     REALM_ASSERT(m_sess);
 
     if (!m_progress_handler)
         return;
 
-    std::uint_fast64_t downloaded_bytes = 0;
-    std::uint_fast64_t downloadable_bytes = 0;
-    std::uint_fast64_t uploaded_bytes = 0;
-    std::uint_fast64_t uploadable_bytes = 0;
-    std::uint_fast64_t snapshot_version = 0;
-    ClientHistory::get_upload_download_bytes(m_db.get(), downloaded_bytes, downloadable_bytes, uploaded_bytes,
-                                             uploadable_bytes, snapshot_version);
+    // Ignore progress messages from before we first receive a DOWNLOAD message
+    if (!m_reliable_download_progress)
+        return;
+
+    ReportedProgress p = m_reported_progress;
+    ClientHistory::get_upload_download_bytes(m_db.get(), p.downloaded, p.downloadable, p.uploaded, p.uploadable,
+                                             p.snapshot);
 
     // If this progress notification was triggered by a commit being made we
     // only want to send it if the uploadable bytes has actually increased,
     // and not if it was an empty commit.
-    if (only_if_new_uploadable_data && m_last_reported_uploadable_bytes == uploadable_bytes)
+    if (only_if_new_uploadable_data && m_reported_progress.uploadable == p.uploadable)
         return;
-    m_last_reported_uploadable_bytes = uploadable_bytes;
 
     // uploadable_bytes is uploaded + remaining to upload, while downloadable_bytes
     // is only the remaining to download. This is confusing, so make them use
     // the same units.
-    std::uint_fast64_t total_bytes = downloaded_bytes + downloadable_bytes;
-
-    m_sess->logger.debug("Progress handler called, downloaded = %1, "
-                         "downloadable(total) = %2, uploaded = %3, "
-                         "uploadable = %4, reliable_download_progress = %5, "
-                         "snapshot version = %6",
-                         downloaded_bytes, total_bytes, uploaded_bytes, uploadable_bytes,
-                         m_reliable_download_progress, snapshot_version);
-
-    // FIXME: Why is this boolean status communicated to the application as
-    // a 64-bit integer? Also, the name `progress_version` is confusing.
-    std::uint_fast64_t progress_version = (m_reliable_download_progress ? 1 : 0);
-    m_progress_handler(downloaded_bytes, total_bytes, uploaded_bytes, uploadable_bytes, progress_version,
-                       snapshot_version);
+    p.downloadable += p.downloaded;
+
+    double upload_estimate = 1.0, download_estimate = 1.0;
+
+    if ((!is_completed || is_download) && p.uploaded != p.uploadable) {
+        if (p.uploadable > p.final_uploaded)
+            upload_estimate = (p.uploaded - p.final_uploaded) / double(p.uploadable - p.final_uploaded);
```

Comment:
why do the diff? shouldn't the estimate be `p.uploaded/p.uploadable`? IIUC, final bytes only make sense for download (where the bootstrap is added)

Reply:
This values are always simply increasing, so if you add 1mb of data you will get values from 0 to 1048576.
Then if you add one more 1mb reported values are going to be 1048576 to 2097152, and so on.
So the progress reported will be always shorter and shorter range first from 0.0 to 1.0, then from 0.5 to 1.0, then 0.66 to 1.0. How does it make sense? That is also not how server sends its progress for download. It always restarts after completion.

Reply:
I think the reason why the progress always increases for download is because the server (cc @rohitn212) associates the remaining changes with a server version, and then it keeps it at 1.0 past that unless there is a change of query. For upload things are a bit different (since we know exactly the changes we need to upload and also there is no bootstrap), so I think it's fine for progress to increase and decrease as changes are committed and acknowledged by the server. Alternatively, we can set the last changeset as the one the progress will report 1.0 for, then decrease and increase towards 1.0 until again the last changeset is uploaded and so on. I am not sure what strategy we follow with these changes.

Reply:
Yes, on the server side, we ensure that the progress estimate remains non-increasing. We also only send the full 100% (1.0) value for the final download changeset.

Reply:
Oops, yes. Sorry, that was a typo. 
The COQ wouldn't be processed until after whatever triggered the initial history scan (ex BindQuery). But yes, once the initial history scan is complete and the client has caught up after reconnecting then the progress will end at 1.0.
The COQ would be a separate instance so the progress starting from 0 doesn't break the non-decreasing rule.

Reply:
interesting. i was not aware a coq is postponed in this case.

Reply:
If the client reconnected with just a Bind query then it wouldn't trigger the initial history scan.
@sudssm can you confirm whether what I said above is correct?

Reply:
@danieltabacaru what's your suggestion for this approach? I've tweaked testing code to check a few different scenarios. There is 'dump' function for progress values reported which is invoked on test failure, but you may simply call in on every 'clear(progress)' call to see which values were reported for every type of progress callback in upload/download pbs/flx. With this logic i saw most consistent behaviour. Do you think it is still unreasonable to proceed this way? Or is it fine to resolve?

Reply:
@rohitn212 that's correct. but this code looks like its upload progress, which we havent made any changes to?
for download progress, the server will always send monotonically increasing values between 0 and 1, for each bootsrap or COQ. COQs are processed in serial, and we will not start one until the previous (bootstrap or COQ) is complete

Reply:
I think we'll have to chat in person about this.

Reply:
I wrote a test to see what's going on here and I think I've figured it out.
```
TEST_CASE("flx: progress debugging", "") {
    FLXSyncTestHarness harness("flx_bootstrap_reconnect", {g_large_array_schema, {"queryable_int_field"}});
    SyncTestFile uploader_config(harness.app()->current_user(), harness.schema(), SyncConfig::FLXSyncEnabled{});

    auto uploader_realm = Realm::get_shared_realm(uploader_config);
    {
        auto mut_subs = uploader_realm->get_latest_subscription_set().make_mutable_copy();
        auto table = uploader_realm->read_group().get_table("class_TopLevel");
        mut_subs.insert_or_assign(Query(table));
        mut_subs.commit().get_state_change_notification(sync::SubscriptionSet::State::Complete).get();
    };

    enum State { Start, FilledInitialData, InNotifier, FilledSecondData };
    TestingStateMachine<State> state(Start);

    auto logger = util::Logger::get_default_logger();
    auto fill_realm = [generation = 1, logger](SharedRealm realm) mutable {
        CppContext c(realm);
        for (int i = 0; i < 5; ++i) {
            realm->begin_transaction();
            auto obj =
                Object::create(c, realm, "TopLevel",
                               std::any(AnyDict{{"_id", ObjectId::gen()},
                                                {"list_of_strings", AnyVector{}},
                                                {"queryable_int_field", static_cast<int64_t>(i + generation)}}));
            List str_list(obj, realm->schema().find("TopLevel")->property_for_name("list_of_strings"));
            for (int j = 0; j < 1024; ++j) {
                str_list.add(c, std::any(std::string(1024, 'a' + (j % 26))));
            }

            realm->commit_transaction();
            logger->info("Committed transaction for object %1", i + generation);
        }
        generation += 100;
    };

    auto progress_nofitier = [&](uint64_t transferred_bytes, uint64_t transferrable_bytes, double) {
        bool should_wait = false;
        state.transition_with([&](State cur_state) -> std::optional<State>{
            if (cur_state == FilledInitialData && transferred_bytes > 0 && transferrable_bytes != transferred_bytes) {
                should_wait = true;
                return InNotifier;
            }
            return std::nullopt;
        });
        if (should_wait) {
            state.wait_for(FilledSecondData);
        }
    };
    uploader_realm->sync_session()->register_progress_notifier(std::move(progress_nofitier),
                                                               SyncSession::ProgressDirection::upload, true);
    fill_realm(uploader_realm);
    state.transition_with([&](State cur_state) {
        REQUIRE(cur_state == Start);
        return FilledInitialData;
    });
    state.wait_for(InNotifier);
    fill_realm(uploader_realm);
    state.transition_with([&](State cur_state) {
        REQUIRE(cur_state == InNotifier);
        return FilledSecondData;
    });
    wait_for_upload(*uploader_realm);
}
```

I think the logic here is that if you don't do this difference calculation between uploadable/uploaded/final_uploaded, when you add a bunch of stuff to upload, the progress estimate doesn't go back to zero, it goes back to some percentage non-zero percentage. This calculation lets you have a clean progression from zero to 1.0 even if you then add more data to upload before it reaches 1.0, and then once you reach 1.0 the next batch of uploads will restart the estimate at zero.

Can we just add a comment to that effect?

Reply:
I ran the test and the progress goes back to zero only after the upload is complete, not if more data is added before progress reaches 1.0.

@jbreams Not sure what you mean by:
> This calculation lets you have a clean progression from zero to 1.0 even if you then add more data to upload before it reaches 1.0

Is it that it always increases? Because that's not the case (try changing transferred_bytes > 0 into transferred_bytes > 78). The progress goes from 0.2 to 0.1 (the same with my proposed calculation).

I think this is in line with what I mentioned here though:
> Alternatively, we can set the last changeset in history as the one the progress will report 1.0 for, then decrease and increase towards 1.0 until again the last changeset is uploaded and so on.

All in all, this has the nice property that it behaves (more or less) similarly with download progress.

Reply:
Sorry, I meant that it goes from zero to 1.0 until everything is uploaded, but when everything is uploaded and you add new data it resets to zero, like this 0 -> 1.0 -> 0.1 -> 0.0 -> 0.1 ... -> 0.4 -> new data added -> 0.3 ... -> 1. Otherwise it goes from 0 -> 1.0 -> 0.0 -> 0.1 ... -> 0.4 -> new data added -> 0.3 ... -> 1.